src/scatter_buildblock/ScatterEstimation.cxx

```diff
@@ -0,0 +1,1486 @@
+        return true;
+    }
+    else // Parse locally
+    {
+        KeyParser local_parser;
+        local_parser.add_start_key("Scatter Simulation");
+        local_parser.add_stop_key("End Scatter Simulation");
+        local_parser.add_parsing_key("Simulation method", &this->scatter_simulation_sptr);
+        if (!local_parser.parse(this->scatter_sim_par_filename.c_str()))
+        {
+            warning(boost::format("ScatterEstimation: Error parsing scatter simulation parameters file %1%. Aborting.")
+                    %this->recon_template_par_filename);
+            return true;
+        }
+    }
+
+    if (this->output_scatter_estimate_prefix.size() == 0)
+        return true;
+
+    if (this->output_background_estimate_prefix.size() == 0)
+        return true;
+
+    if(!this->recompute_mask_projdata)
+    {
+        if (this->mask_projdata_filename.size() == 0)
+        {
+            warning("ScatterEstimation: Please define a filename for mask proj_data. Aborting.");
+            return true;
+        }
+        this->mask_projdata_sptr =
+                ProjData::read_from_file(this->mask_projdata_filename);
+    }
+    else
+    {
+        if (!this->recompute_mask_image)
+        {
+            if (this->mask_image_filename.size() == 0 )
+            {
+                warning("ScatterEstimation: Please define a filename for mask image. Aborting.");
+                return true;
+            }
+
+            this->mask_image_sptr =
+                    read_from_file<DiscretisedDensity<3, float> >(this->mask_image_filename);
+        }
+
+        if (this->tail_mask_par_filename.size() == 0)
+        {
+            warning("ScatterEstimation: Please define a filename for tails mask. Aborting.");
+            return true;
+        }
+    }
+
+    return false;
+}
+
+Succeeded
+ScatterEstimation::
+set_up()
+{
+    if (this->run_debug_mode)
+    {
+        info("ScatterEstimation: Debugging mode is activated.");
+        this->export_scatter_estimates_of_each_iteration = true;
+
+        // Create extras folder in this location
+        FilePath current_full_path(FilePath::get_current_working_directory());
+        extras_path = current_full_path.append("extras");
+    }
+
+    if (is_null_ptr(this->input_projdata_sptr))
+    {
+        warning("ScatterEstimation: No input proj_data have been set. Aborting.");
+        return Succeeded::no;
+    }
+
+#if 0
+    // Calculate the SSRB
+    if (input_projdata_sptr->get_num_segments() > 1)
+    {
+        info("ScatterEstimation: Running SSRB on input data...");
+        shared_ptr<ProjDataInfo> proj_data_info_2d_sptr(
+                    dynamic_cast<ProjDataInfoCylindricalNoArcCorr* >
+                    (SSRB(*this->input_projdata_sptr->get_proj_data_info_ptr(),
+                          this->input_projdata_sptr->get_num_segments(), 1, false)));
+
+        FilePath tmp(this->input_projdata_filename);
```

Comment:
in the future (or maybe already from SIRF?), we won't necessarily have an input filename. let's just mark this with a `TODO` comment

Reply:
ok